src/opcode.rs

```diff
@@ -1,45 +1,420 @@
+        assert_eq!(opcode, Opcode::OP_LDI { dr: 1, offset: 1 });
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_ldr() -> Result<(), OpcodeError> {
+        let instruction = 0b0110_0010_1000_0001;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(
+            opcode,
+            Opcode::OP_LDR {
+                dr: 1,
+                base: 2,
+                offset: 1
+            }
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_lea() -> Result<(), OpcodeError> {
+        let instruction = 0b1110_0010_0000_0001;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(opcode, Opcode::OP_LEA { dr: 1, offset: 1 });
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_not() -> Result<(), OpcodeError> {
+        let instruction = 0b1001_0010_1011_1111;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(opcode, Opcode::OP_NOT { dr: 1, sr: 2 });
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_rti() -> Result<(), OpcodeError> {
+        let instruction = 0b1000_0000_0000_0000;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(opcode, Opcode::OP_RTI);
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_st() -> Result<(), OpcodeError> {
+        let instruction = 0b0011_0010_0000_0001;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(opcode, Opcode::OP_ST { sr: 1, offset: 1 });
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_sti() -> Result<(), OpcodeError> {
+        let instruction = 0b1011_0010_0000_0001;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(opcode, Opcode::OP_STI { sr: 1, offset: 1 });
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_str() -> Result<(), OpcodeError> {
+        let instruction = 0b0111_0010_1000_0001;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(
+            opcode,
+            Opcode::OP_STR {
+                sr: 1,
+                base: 2,
+                offset: 1
+            }
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_trap() -> Result<(), OpcodeError> {
+        let instruction = 0b1111_0000_0000_0011;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(opcode, Opcode::OP_TRAP { trapvec: 3 });
+        Ok(())
+    }
+
+    #[test]
+    fn test_op_res() -> Result<(), OpcodeError> {
+        let instruction = 0b1101_0000_0000_0000;
+        let opcode = Opcode::from(instruction)?;
+        assert_eq!(opcode, Opcode::OP_RES);
+        Ok(())
+    }
+
+    #[test]
+    fn test_sign_extend_positive() {
+        let value = 0b00011;
+        let bit_count = 5;
+        let result = sign_extend(value, bit_count);
+        assert_eq!(result, 0b00011);
+    }
+
+    #[test]
+    fn test_sign_extend_negative() {
+        let value = 0b11111;
+        let bit_count = 5;
+        let result = sign_extend(value, bit_count);
+        assert_eq!(result, 0xFFFF);
+    }
```

Comment:
Very minor nitpick, but maybe I would some test(s) to check what happens when you fail to decode an instruction. 


Things like gibberish and also "similar codes". For instance, testing what happens when you try to decode:
`0b1111_0000_0000_0111` which is _similar_ to OP_TRAP.